### PR TITLE
fix(Notification): prevent private item exposure from notification

### PR DESCRIPTION
**Pull request by stonebuzz**
When a ```requester``` have 2 profiles
 - ```self-service```
 - ```tech``` (without see private followup)

Notification contain private followup

why ?

When ```NotificationTargetCommonITILObject``` call ``` $item->getTimelineItems($options);```

With this options
```
            $options = [
                'with_documents'    => false,
                'with_logs'         => false,
                'with_validations'  => false,
                'expose_private'    => $show_private,
                'bypass_rights'     => true,
                'sort_by_date_desc' => true,
                'is_self_service'   => $is_self_service,
            ];
```

```$show_private``` is ```false``` because it's computed according to the profiles that have the right to ``ITILFollowup::SEEPRIVATE``.

```$is_self_service``` is ```false```, because ```requester``` have one profile with ```interface``` = ```central``` (tech)

First option ```$show_private``` is used with current session 

```php
 $restrict_fup = $restrict_task = [];
        if (!$params['expose_private'] && !Session::haveRight("followup", ITILFollowup::SEEPRIVATE)) {
            $restrict_fup = [
                'OR' => [
                    'is_private' => 0,
                    'users_id'   => Session::getLoginUserID()
                ]
            ];
        }
```

while the second option is always used ( ```helpdesk``` interface, notification etc ...)

```php
        if ($params['is_self_service'] ) {
            $restrict_fup = [
                'is_private'   => 0
            ];
        }
```

In my case all private followup are send.

this PR prevent this

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !27671

### CODE REVIEW

File: src/CommonITILObject.php [hunk 1]
```diff
@@ -6759,7 +6759,7 @@ public function getTimelineItems(array $options = [])
             ];
         }
 
-        if ($params['is_self_service']) {
+        if ($params['is_self_service'] || (!$params['expose_private'] &&  $params['bypass_rights'])) {
```
Comment (cedric-anne) [inline src/CommonITILObject.php:6762]: This new condition is a bit hard to read.

I think the fix has to be made on previous condition:
```
- if (!$params['expose_private'] && !Session::haveRight("followup", ITILFollowup::SEEPRIVATE)) {
+if (!$params['expose_private'] || (!Session::haveRight("followup", ITILFollowup::SEEPRIVATE) && !$params['bypass_rights'])) { 
```

Comment (stonebuzz) [reply src/CommonITILObject.php]: it's work too

File: tests/functional/NotificationTargetTicket.php [hunk 2]
```diff
@@ -360,6 +360,52 @@ public function testTimelineTag()
...
+        //add second test for post-only
+        //simulate that is a self-service (because of one profil with interface = helpdesk)
+        //he can't see private because of self-service who has no right TILFollowup::SEEPRIVATE
+        $basic_options = [
+            'additionnaloption' => [
+                'usertype' => NotificationTarget::GLPI_USER,
+                'is_self_service' => true,
+                'show_private'    => false,
+            ]
+        ];
```
Comment (cedric-anne) [inline tests/functional/NotificationTargetTicket.php:372]: This is exactly the same test as the previous one.

Comment (stonebuzz) [reply tests/functional/NotificationTargetTicket.php]: fixed

**Comment by stonebuzz** [comment]
I came back to the previous test

```php
  if ($params['is_self_service'] || !$params['expose_private']) {
      $restrict_task = [
          'is_private'   => 0
      ];
  }
```

Check ```$params['bypass_rights']``` here  is not enought

```php
        if (!$params['expose_private'] || (!Session::haveRight("followup", ITILFollowup::SEEPRIVATE) && !$params['bypass_rights'])) {
            $restrict_fup = [
                'OR' => [
                    'is_private' => 0,
                    'users_id'   => Session::getLoginUserID()
                ]
            ];
        }
```

Private items are load

File: src/NotificationTargetCommonITILObject.php [hunk 3]
```diff
@@ -1591,6 +1591,8 @@ public function getDataForObject(CommonDBTM $item, array $options, $simple = fal
                 'is_self_service'   => $is_self_service,
             ];
 
+            //Toolbox::logDebug($show_private, $is_self_service);
+
```
Comment (cedric-anne) [inline src/NotificationTargetCommonITILObject.php:1595]:
```suggestion
```

File: tests/functional/NotificationTargetTicket.php [hunk 4]
```diff
@@ -360,6 +360,54 @@ public function testTimelineTag()
             ]
         ];
 
+        //add second test for post-only
+        //simulate that is a not self-service (because of one profil with interface = central)
+        //he can't see private because of self-service who has no right TILFollowup::SEEPRIVATE
+        //log as tech to trigger Session check (As if tech was at the origin of noticication )
```
Comment (cedric-anne) [inline tests/functional/NotificationTargetTicket.php:366]:
```suggestion
        //add a test for tech, but force the `show_private` option to false to ensure that presence of this option will
        //hide private items
```

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**